File: 2020/knowit/2/2.js
"use strict";

const erathostenes = n => {
	const primes = Array(n+1).fill(true);
	primes[0] = false;
	primes[1] = false;

	let k = 2;
	const sqrt = Math.sqrt(n);
	while (k <= sqrt) {
		for (let i = 2; i*k <= n; i++) {
			primes[i*k] = false;
		}

		do {
			k++;
		} while (!primes[k]);
	}

	return primes;
}

const luke2 = n => {
	const primes = erathostenes(n);

	let i = 0;
	let delivered = 0;
	while (i < n) {
		if (String(i).includes("7")) {
			let p = i;
			while (!primes[p]) {
				p--;
			}

			i += p;
		} else {
			delivered++;
		}

		i++;
	}

	return delivered;
}

console.log(luke2(5433000));

File: 2020/knowit/2/2.hs
import qualified Data.Set as S

-- Stolen from https://www.reddit.com/r/haskell/comments/35vc31/the_real_way_to_generate_a_list_of_primes_in/
primes :: [Int]
primes = 2 : primes'
    where isPrime (p:ps) n = p*p > n || n `rem` p /= 0 && isPrime ps n
          primes' = 3 : filter (isPrime primes') [5, 7 ..]

primeSet :: Int -> S.Set Int
primeSet n =
    S.fromDistinctAscList
    . takeWhile (<= n)
    $ primes

includes7 :: Int -> Bool
includes7 0 = False
includes7 n = n `mod` 10 == 7 || includes7 (n `div` 10)

luke2 :: Int -> Int
luke2 n =
    luke2' 0 0
    where luke2' i delivered
              | i >= n      = delivered
              | includes7 i = luke2' (i+(p i)+1) delivered
              | otherwise   = luke2' (i+1) (delivered+1)
          p i
              | S.member i ps = i
              | otherwise     = p (i-1)
          ps = primeSet n

main = print . luke2 $ 5433000
